tap/src/dot11/nl.rs

```diff
@@ -74,178 +41,151 @@ pub enum InterfaceState {
-                            let start = freq_cursor;
-                            let freq_length = LittleEndian::read_u16(&freqs[freq_cursor..freq_cursor+2]);
-                            freq_cursor = start+freq_length as usize;
-
-                            if let Some(f) = parse_frequency_from_attributes(&freqs[start..start+freq_length as usize]) {
-                                supported_frequencies.push(f);
-                            }
+    let mut phy_index: Option<u32> = None;
+    let mut supported_frequencies: Vec<u32> = Vec::new();
 
-                            if length == (freq_cursor+4) as u16 {
-                                break;
+    for msg in msgs {
+        if let NlPayload::Payload(payload) = msg.nl_payload() {
+            for attr in payload.attrs().iter() {
+                match attr.nla_type().nla_type() {
+                    &Nl80211Attribute::WiPhy => {
+                        phy_index = Some(attr.get_payload_as().map_err(|_| DeError::new("Failed to get payload as u32"))?);
+                    },
+                    &Nl80211Attribute::WiPhyBands => {
+                        // bands is an array, get the handle for the array
+                        let bands_handle = attr.get_attr_handle::<Nl80211BandAttr>().map_err(|_| DeError::new("Failed to get WiPhyBands attribute handle"))?;
+                        for bands in bands_handle.iter() {
+                            // for each element of the array now we get the nested attributes
+                            let band_handle = bands.get_attr_handle().map_err(|_| DeError::new("Failed to get WiPhyBands attribute handle"))?;
+
+                            for band_attr in band_handle.get_attrs() {
+                                match band_attr.nla_type().nla_type() {
+                                    &Nl80211BandAttr::Freqs => {
+                                        // freqs is an array, get the handle for the array
+                                        let freqs_handle = band_attr.get_attr_handle::<Nl80211FrequencyAttr>().map_err(|_| DeError::new("Failed to get WiPhyFreq attribute handle"))?;
+                                        for freq in freqs_handle.iter() {
+                                            // for each element of the array now we get the nested attributes
+                                            let freq_handle = freq.get_attr_handle().map_err(|_| DeError::new("Failed to get WiPhyFreq attribute handle"))?;
+                                            let freq_attr = freq_handle.get_attribute(Nl80211FrequencyAttr::Freq);
+                                            let disabled_attr = freq_handle.get_attribute(Nl80211FrequencyAttr::Disabled);
+
+                                            if freq_attr.is_some() && disabled_attr.is_none() {
+                                                supported_frequencies.push(freq_attr.unwrap().get_payload_as()?);
+                                            }
```

Comment:
Same here for the `unwrap()` IMO

Reply:
We agreed that unwrap is ok when used in specific situations where a panic would mean a bug.